recipe/session/errors/errors.go

```diff
@@ -0,0 +1,91 @@
+package errors
+
+import (
+	"reflect"
+)
+
+const (
+	UnauthorizedErrorStr       = "UNAUTHORISED"
+	TryRefreshTokenErrorStr    = "TRY_REFRESH_TOKEN"
+	TokenTheftDetectedErrorStr = "TOKEN_THEFT_DETECTED"
+)
+
+type SessionError struct{}
+
+// TryRefreshTokenError used for when the refresh API needs to be called
+type TryRefreshTokenError struct {
+	Msg  string
+	Type string
+}
+
+// TODO: Do we need this function? Some places we are not using it.
+func MakeTryRefreshTokenError(msg string) TryRefreshTokenError {
+	return TryRefreshTokenError{
+		Msg: msg,
+		// TODO: Do we need this Type field?
+		Type: TryRefreshTokenErrorStr,
+	}
+}
+
+func (err TryRefreshTokenError) Error() string {
+	return err.Msg
+}
+
+// TokenTheftDetectedError used for when token theft has happened for a session
+type TokenTheftDetectedError struct {
+	Msg     string
+	Type    string
+	Payload TokenTheftDetectedErrorPayload
+}
+
+type TokenTheftDetectedErrorPayload struct {
+	SessionHandle string
+	UserID        string
+}
+
+func MakeTokenTheftDetectedError(sessionHandle, userID, msg string) TokenTheftDetectedError {
+	return TokenTheftDetectedError{
+		Msg:  msg,
+		Type: TokenTheftDetectedErrorStr,
+		Payload: TokenTheftDetectedErrorPayload{
+			SessionHandle: sessionHandle,
+			UserID:        userID,
+		},
+	}
+}
+
+func (err TokenTheftDetectedError) Error() string {
+	return err.Msg
+}
+
+// UnauthorizedError used for when the user has been logged out
+type UnauthorizedError struct {
+	Msg  string
+	Type string
+}
+
+func MakeUnauthorizedError(msg string) UnauthorizedError {
+	return UnauthorizedError{
+		Msg:  msg,
+		Type: UnauthorizedErrorStr,
+	}
+}
+
+func (err UnauthorizedError) Error() string {
+	return err.Msg
+}
+
+// IsTokenTheftDetectedError returns true if error is a TokenTheftDetectedError
+func IsTokenTheftDetectedError(err error) bool {
```

Comment:
Hi,

is there a particular reason why don't use the `errors.As` method?

Reply:
errors.IsTokenTheftDetectedError makes more sense than error.AsTokenTheftDetectedError.

Reply:
sorry, I meant Go's native method for errors comparisons https://blog.golang.org/go1.13-errors#TOC_3.2.

Reply:
great suggestion! thanks for pointing this out :)